Haskell/Haskell.sublime-syntax

```diff
@@ -5,68 +5,142 @@ name: Haskell
 file_extensions:
   - hs
 scope: source.haskell
+
+variables:
+  # In case this regex seems overly general,
+  # note that Haskell permits the definition of new operators
+  # which can be nearly any string of punctuation characters,
+  # such as $%^&*.
+  operator_char: '[*|!%&$?~+:\-.=</>\\]'
+  operator_infix: '{{operator_char}}+'
+  operator_parens: '(?:{{operator_infix}}|,+)'
+  module_name: (?:[A-Z][A-Za-z._']*)
+  escape_chars: |-
+    (?x:NUL|SOH|STX|ETX|EOT|ENQ|ACK|BEL|BS|HT|LF|VT|FF|CR|SO|SI
+      |DLE|DC1|DC2|DC3|DC4|NAK|SYN|ETB|CAN|EM|SUB|ESC|FS|GS|RS|US|SP|DEL
+      |[abfnrtv\\\"'\&])
+  escape_sequence: |-
+    (?x:
+      (\\{{escape_chars}})   # Escapes
+      | (\\[0-9]+)           # Decimal Escapes
+      | (\\o[0-7]+)          # Octal Escapes
+      | (\\x[0-9A-Fa-f]+)    # Hexadecimal Escapes
+      | (\^[A-Z@\[\]\\\^_])  # Control Chars
+    )
+
 contexts:
+  prototype:
+    - include: comment
+    - include: preprocessor
+  
   main:
-    - match: "(`)[a-zA-Z_']*?(`)"
-      comment: "In case this regex seems unusual for an infix operator, note that Haskell allows any ordinary function application (elem 4 [1..10]) to be rewritten as an infix expression (4 `elem` [1..10])."
-      scope: keyword.operator.function.infix.haskell
-      captures:
-        1: punctuation.definition.entity.haskell
-        2: punctuation.definition.entity.haskell
+    - include: statement
+    - include: expression
+
+  statement:
+    - include: declaration
+    - include: import
+
+  expression:
+    - include: keyword
+    - include: operator
+    - include: constant
+    - include: string
+    - include: splice
+    - include: group
+    - include: list
+    - include: ident
+    - match: ','
+      scope: punctuation.separator.comma.haskell
+
+  group:
+    - match: \(
+      scope: punctuation.section.group.begin.haskell
+      push:
+        - meta_scope: meta.group.haskell
```

Comment:
Do we really need/care about this one? On the screenshot below (in the middle), it matches `round` but it shouldn't.

![image](https://user-images.githubusercontent.com/287532/74486473-4fa1a300-4ebd-11ea-935e-b3539151741b.png)

Reply:
Yes, we care about groups in general and mark the entire thing with a meta scope to signify its spanning tree, sort of. We do this for all syntaxes, so not doing it for Haskell would be more than odd.

This sounds like an XY problem to me. What is the behavior you expect (for a given code sample) and what happens instead?

Reply:
I'd expect that `round` wouldn't match with `meta.group.haskell` :thinking:

Reply:
The prelude not being matched inside the group is a separate issue and while the removal of these matches work, it's not an actual fix for the issue. I absolutely won't remove these matches and instead work on a proper fix for matching items of the prelude in groups, which is what I interpreted your report as.

Reply:
>and instead work on a proper fix for matching items of the prelude in groups, which is what I interpreted your report as.

:100: